Feat/generate and store feedback new structure

OhMyMy426 opened the pull request:
closed #236 
This pull request introduces several enhancements and new features to the backend application, focusing on improving training case creation, preparation, and feedback processes. Key changes include refactoring imports for better organization, adding background task handling for training preparation, introducing new schemas and services, and implementing unit tests for the training case service.

### Refactoring and Code Organization:
* [`backend/app/models/__init__.py`](diffhunk://#diff-d5870c67917db2ab2c21aa4a6553eaf5b9b2d105f1c46623461fff72e7c59ae6L1-R56): Refactored imports to use absolute paths, ensuring better maintainability and clarity. Added `FeedbackStatusEnum` to the `__all__` list for export. [[1]](diffhunk://#diff-d5870c67917db2ab2c21aa4a6553eaf5b9b2d105f1c46623461fff72e7c59ae6L1-R56) [[2]](diffhunk://#diff-d5870c67917db2ab2c21aa4a6553eaf5b9b2d105f1c46623461fff72e7c59ae6R104)

### Training Case Enhancements:
* [`backend/app/routers/training_case_route.py`](diffhunk://#diff-1d20022e1ebc768b0b2bae23fa8ac77d6373fe960446d060b860fa10d998bdb6L31-R79): Updated the `create_training_case` endpoint to include background task handling for training preparation. Added validation of foreign keys, initialization of preparation records, and a response indicating the preparation process has started.
* [`backend/app/services/training_case_service.py`](diffhunk://#diff-fdecc24c9369f3b3db0bad241a249ad8c0cb4049a7e4986e0baac0ceda200ae0R1-R14): Added a new service method `create_training_case` to encapsulate the logic for creating a training case in the database.

### Training Preparation Improvements:
* [`backend/app/schemas/training_preparation_schema.py`](diffhunk://#diff-e180971a168579d40c4b11bb2606ce6da3a097e001eb35a3ab156482cae406a0R34-R49): Introduced new schemas `TrainingPreparationRequest` and `TrainingPreparationResponse` for handling training preparation requests and responses.
* [`backend/app/services/training_preparation_service.py`](diffhunk://#diff-d77bc1b54457ec4978f211a4e8573d1287cbd0fe9a92d2823b48510d1a5f82f5R172-R263): Added methods `create_pending_preparation` and `generate_training_preparation` to handle the initialization and generation of training preparation data, including objectives, checklists, and key concepts.

### Training Feedback Enhancements:
* [`backend/app/services/training_feedback_service.py`](diffhunk://#diff-e76dda7293dc1149f926e29349f2054083f93ccb9b7ff58114a8ca248e4ea684R228-R281): Added a new method `generate_and_store_feedback` to generate feedback based on session data and store it in the database.

### Unit Testing:
* [`backend/app/tests/services/test_training_case_service_db.py`](diffhunk://#diff-3451b7f34329d6fd8b9fffef71e70e1f8b6c6ace47b80ccba0343223e7e002a9R1-R60): Implemented unit tests for the `create_training_case` service method to ensure correctness and reliability.

rautenrieb commented on backend/app/services/training_feedback_service.py:
```diff
@@ -219,6 +225,60 @@ def generate_recommendations(request: RecommendationsRequest) -> Recommendations
+    goals = get_achieved_goals(goals_request)
+    recommendations = generate_recommendations(recommendations_request)
+
+    examples_positive_dicts = [ex.dict() for ex in examples.positive_examples]
```
For the next three lines: `obj.dict()` was deprecated in SQLModel 0.0.14, you should instead use `obj.model_dump()`.

OhMyMy426 replied:
Oh yeah, it is the same issue XD

rautenrieb reviewed (changes requested):
This is looks awesome! Thanks Xiyu. See my small comment about the .dict()

OhMyMy426 commented:
> Hi Xiyu, see comment.
> 
> Also, what happens if examples_positive_dicts, examples_negative_dicts, recommendations fails? Should we add a retry as well here? What do you think?

Thanks for the tipp, yeah I have also updated the same retry logic as in the training_preparation and also added test cases
